[XLib] Add Setting Window Min/Max, Fix Sizeable [Win32] Fix Maximize

time-killer-games opened the pull request:
Fixes the window_set_sizeable() to actually work, and the GML functions added since GMS 1.x - window_set_min/max_width()/height(). https://docs.yoyogames.com/source/dadiospice/002_reference/windows%20and%20views/the%20game%20window/window_set_max_width.html

Edit: this also fixes fullscreen and resizeable window bugs with the graphics not rendering scaled to the window size. All of that is no longer an issue! What can I say other than, "you're welcome!"

Edit2: Also fixed a bug with Win32 when trying to maximize the window, scaling is now handled correctly here, not just in xlib.

Edit3: also fixes get/set fullscreen switching. Global Game Settings for setting fullscreen, borderless mode, and resizeable window on both Win32 and XLib. You can now use dialogs in fullscreen with all graphics systems and platforms. Fixes borderless switching on Windows and setting to borderless on XLib. Fixes sizable switching on Wind32 and XLib.

Here's a demo linux exe (x64) and gmk that demonstrates a proper borderless toggle: [game.zip](https://github.com/enigma-dev/enigma-dev/files/3484633/game.zip)

RobertBColton commented on ENIGMAsystem/SHELL/Platforms/General/PFwindow.h:
```diff
@@ -237,6 +237,11 @@ int window_get_region_height();
+void window_set_min_width(int width);
+void window_set_min_height(int height);
+void window_set_max_width(int width);
+void window_set_max_height(int height);
```
Maybe we want to move these to General so that the game has them for all platfoms. We're going to want to add this to Win32 anyway with WM_GETMINMAXINFO message. I think we'll need the state there anyway, not sure.

RobertBColton commented on ENIGMAsystem/SHELL/Platforms/xlib/XLIBwindow.cpp:
```diff
@@ -43,6 +44,11 @@
+static int window_get_min_width = -1;
+static int window_get_max_width = -1;
+static int window_get_min_height = -1;
+static int window_get_max_height = -1;
```
Maybe we want to move these to General so that the game has them for all platfoms. We're going to want to add this to Win32 anyway with WM_GETMINMAXINFO message. I think we'll need the state there anyway, not sure.

time-killer-games replied:
I can't put it in General becuase it produces the errors in my comment farther down this page.

RobertBColton replied:
You got this one, this is resolved.

RobertBColton commented:
It looks pretty good, I can close #1341, thanks for working on this one. After we get this merged I can add it for Win32 and SDL or you can if you want to.

time-killer-games commented:
@RobertBColton wtf

```
Linking /tmp/egm2071147129716961429.tmp
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/Platforms/General/PFwindow.o:(.data+0x4): multiple definition of `enigma::window_max_height'
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/SHELLmain.o:(.data+0x40): first defined here
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/Platforms/General/PFwindow.o:(.data+0x8): multiple definition of `enigma::window_min_height'
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/SHELLmain.o:(.data+0x44): first defined here
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/Platforms/General/PFwindow.o:(.data+0xc): multiple definition of `enigma::window_max_width'
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/SHELLmain.o:(.data+0x48): first defined here
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/Platforms/General/PFwindow.o:(.data+0x10): multiple definition of `enigma::window_min_width'
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/SHELLmain.o:(.data+0x4c): first defined here
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/Platforms/General/PFmain.o:(.data+0x4): multiple definition of `enigma::window_max_height'
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/SHELLmain.o:(.data+0x40): first defined here
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/Platforms/General/PFmain.o:(.data+0x8): multiple definition of `enigma::window_min_height'
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/SHELLmain.o:(.data+0x44): first defined here
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/Platforms/General/PFmain.o:(.data+0xc): multiple definition of `enigma::window_max_width'
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/SHELLmain.o:(.data+0x48): first defined here
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/Platforms/General/PFmain.o:(.data+0x10): multiple definition of `enigma::window_min_width'
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/SHELLmain.o:(.data+0x4c): first defined here
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/Platforms/xlib/XLIBwindow.o:(.data+0x3c): multiple definition of `enigma::window_min_width'
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/SHELLmain.o:(.data+0x4c): first defined here
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/Platforms/xlib/XLIBwindow.o:(.data+0x38): multiple definition of `enigma::window_max_width'
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/SHELLmain.o:(.data+0x48): first defined here
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/Platforms/xlib/XLIBwindow.o:(.data+0x34): multiple definition of `enigma::window_min_height'
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/SHELLmain.o:(.data+0x44): first defined here
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/Platforms/xlib/XLIBwindow.o:(.data+0x30): multiple definition of `enigma::window_max_height'
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/SHELLmain.o:(.data+0x40): first defined here
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/Platforms/xlib/XLIBmain.o:(.data+0x0): multiple definition of `enigma::window_max_height'
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/SHELLmain.o:(.data+0x40): first defined here
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/Platforms/xlib/XLIBmain.o:(.data+0x4): multiple definition of `enigma::window_min_height'
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/SHELLmain.o:(.data+0x44): first defined here
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/Platforms/xlib/XLIBmain.o:(.data+0x8): multiple definition of `enigma::window_max_width'
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/SHELLmain.o:(.data+0x48): first defined here
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/Platforms/xlib/XLIBmain.o:(.data+0xc): multiple definition of `enigma::window_min_width'
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/SHELLmain.o:(.data+0x4c): first defined here
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/Platforms/General/POSIX/shell.o:(.data+0x0): multiple definition of `enigma::window_max_height'
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/SHELLmain.o:(.data+0x40): first defined here
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/Platforms/General/POSIX/shell.o:(.data+0x4): multiple definition of `enigma::window_min_height'
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/SHELLmain.o:(.data+0x44): first defined here
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/Platforms/General/POSIX/shell.o:(.data+0x8): multiple definition of `enigma::window_max_width'
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/SHELLmain.o:(.data+0x48): first defined here
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/Platforms/General/POSIX/shell.o:(.data+0xc): multiple definition of `enigma::window_min_width'
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/SHELLmain.o:(.data+0x4c): first defined here
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/Platforms/General/POSIX/POSIXdirectory.o:(.data+0x0): multiple definition of `enigma::window_max_height'
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/SHELLmain.o:(.data+0x40): first defined here
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/Platforms/General/POSIX/POSIXdirectory.o:(.data+0x4): multiple definition of `enigma::window_min_height'
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/SHELLmain.o:(.data+0x44): first defined here
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/Platforms/General/POSIX/POSIXdirectory.o:(.data+0x8): multiple definition of `enigma::window_max_width'
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/SHELLmain.o:(.data+0x48): first defined here
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/Platforms/General/POSIX/POSIXdirectory.o:(.data+0xc): multiple definition of `enigma::window_min_width'
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/SHELLmain.o:(.data+0x4c): first defined here
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/Platforms/General/POSIX/Utility.o:(.data+0x0): multiple definition of `enigma::window_max_height'
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/SHELLmain.o:(.data+0x40): first defined here
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/Platforms/General/POSIX/Utility.o:(.data+0x4): multiple definition of `enigma::window_min_height'
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/SHELLmain.o:(.data+0x44): first defined here
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/Platforms/General/POSIX/Utility.o:(.data+0x8): multiple definition of `enigma::window_max_width'
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/SHELLmain.o:(.data+0x48): first defined here
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/Platforms/General/POSIX/Utility.o:(.data+0xc): multiple definition of `enigma::window_min_width'
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/SHELLmain.o:(.data+0x4c): first defined here
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/Platforms/General/POSIX/main.o:(.data+0x0): multiple definition of `enigma::window_max_height'
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/SHELLmain.o:(.data+0x40): first defined here
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/Platforms/General/POSIX/main.o:(.data+0x4): multiple definition of `enigma::window_min_height'
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/SHELLmain.o:(.data+0x44): first defined here
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/Platforms/General/POSIX/main.o:(.data+0x8): multiple definition of `enigma::window_max_width'
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/SHELLmain.o:(.data+0x48): first defined here
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/Platforms/General/POSIX/main.o:(.data+0xc): multiple definition of `enigma::window_min_width'
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/SHELLmain.o:(.data+0x4c): first defined here
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/Widget_Systems/xlib/zenity.o:(.data+0x0): multiple definition of `enigma::window_max_height'
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/SHELLmain.o:(.data+0x40): first defined here
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/Widget_Systems/xlib/zenity.o:(.data+0x4): multiple definition of `enigma::window_min_height'
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/SHELLmain.o:(.data+0x44): first defined here
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/Widget_Systems/xlib/zenity.o:(.data+0x8): multiple definition of `enigma::window_max_width'
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/SHELLmain.o:(.data+0x48): first defined here
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/Widget_Systems/xlib/zenity.o:(.data+0xc): multiple definition of `enigma::window_min_width'
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/SHELLmain.o:(.data+0x4c): first defined here
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/Widget_Systems/xlib/kdialog.o:(.data+0x0): multiple definition of `enigma::window_max_height'
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/SHELLmain.o:(.data+0x40): first defined here
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/Widget_Systems/xlib/kdialog.o:(.data+0x4): multiple definition of `enigma::window_min_height'
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/SHELLmain.o:(.data+0x44): first defined here
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/Widget_Systems/xlib/kdialog.o:(.data+0x8): multiple definition of `enigma::window_max_width'
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/SHELLmain.o:(.data+0x48): first defined here
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/Widget_Systems/xlib/kdialog.o:(.data+0xc): multiple definition of `enigma::window_min_width'
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/SHELLmain.o:(.data+0x4c): first defined here
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/Universal_System/roomsystem.o:(.data+0x18): multiple definition of `enigma::window_max_height'
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/SHELLmain.o:(.data+0x40): first defined here
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/Universal_System/roomsystem.o:(.data+0x1c): multiple definition of `enigma::window_min_height'
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/SHELLmain.o:(.data+0x44): first defined here
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/Universal_System/roomsystem.o:(.data+0x20): multiple definition of `enigma::window_max_width'
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/SHELLmain.o:(.data+0x48): first defined here
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/Universal_System/roomsystem.o:(.data+0x24): multiple definition of `enigma::window_min_width'
/home/owner/.enigma/.eobjs/Linux/Linux/GNU_GCC_G++/Run/SHELLmain.o:(.data+0x4c): first defined here
collect2: error: ld returned 1 exit status
Makefile:168: recipe for target 'compile_game' failed
make[1]: *** [compile_game] Error 1
make[1]: Leaving directory '/home/owner/ENIGMA/enigma-dev/ENIGMAsystem/SHELL'
Makefile:21: recipe for target 'Game' failed
make: *** [Game] Error 2
```

RobertBColton commented:
@time-killer-games `extern` in the header, only put the definition in just one source.

time-killer-games commented:
@RobertBColton alright, now I'm not only using them once just like before, but also using extern. Same errors.

RobertBColton commented:
Only put the definition and initialization in a single cpp source file. Do not put the definition and initialization in a header, ever. Only put the extern declaration in the header.

time-killer-games commented:
ok, i fixed it.

RobertBColton commented:
If you want I can go ahead and help you tackle Win32 right now. If so, move the setters to that same general source file, but leave window_set_sizeable where you have it. Then, in the Win32 system, add the handler for WM_GETMINMAXINFO message.
https://github.com/enigma-dev/enigma-dev/blob/affd4b7e5c204d6da386a01cd56de2b5f7260725/ENIGMAsystem/SHELL/Platforms/Win32/WINDOWScallback.cpp#L67

We don't have to make Win32 work completely before merging this, but me and you already know Win32 is going to be implemented pretty much the same as Xlib. Therefore, we may as well just move the setters to general now at least.

time-killer-games commented:
@RobertBColton I'd prefer to get this merged and work on Windows a different day.

time-killer-games commented:
@RobertBColton all checks have passed. Please merge it, if that's ok.

RobertBColton commented on ENIGMAsystem/SHELL/Platforms/xlib/XLIBwindow.cpp:
```diff
@@ -267,15 +268,50 @@ bool window_get_stayontop() {
+    sh->min_width = sh->max_width = window_get_width();
+    sh->min_height = sh->max_height = window_get_height();
+  }
+  XSetWMSizeHints(disp, win, sh, XA_WM_NORMAL_HINTS);
```
After digging further about this myself, you could have actually just replaced this with XSetWMNormalHints which is what you need in the most cases. Consider doing that if it saves some lines.
https://stackoverflow.com/a/55837317

time-killer-games replied:
i don't see how that is saving any lines. It's replacing 1 line with 1 line. I'm fine with doing it, I just don't see the difference if both reportedly do the same thing and the number of lines remains unaffected.

RobertBColton replied:
The one I linked is actually what you already have. If you call XSetWMNormalHints, you don't need to construct/initialize that XAtom if I understand correctly. Look more closely.

RobertBColton commented on ENIGMAsystem/SHELL/Platforms/xlib/XLIBwindow.cpp:
```diff
@@ -267,15 +268,50 @@ bool window_get_stayontop() {
 }
 
 void window_set_sizeable(bool sizeable) {
-  if (enigma::isSizeable == sizeable) return;
```
This here is the questionable part of why I am not sure to approve it without asking Josh. Rather than eliminate this redundancy check, maybe it would be better to gut your code below into an internal engine function `enigma::window_update_size_hints` above this line in the source. Then update all of the setters to call the internal engine function. Generally, it doesn't make sense to proceed the sizeable change if the window is already sizeable etc.

RobertBColton replied:
Another questionable thing about this is that you are always setting `PMinSize` and `PMaxSize` even when only one of them is changed. That may not be slow or have any side effects, but the window manager does have these separate for a reason. Maybe check if the parameters are `-1` or different from the previous state and only set the hint that's actually changing. We do this in the display functions too iirc.

time-killer-games replied:
@JoshDreamland

JoshDreamland replied:
As discussed on Discord, this should be fine the way it is. I don't want to forbid an update because our window state cache went stale. Linux desktop environments tend to do whatever they want with your window, at the user's request. You full-screen it, they'll un-full-screen it. You pin it, they'll unpin it. Wouldn't surprise me if setting a window to be resizable was a first-class feature of some DE.

RobertBColton commented:
Alright, it's good but I would like to hear first what Josh says about state in my above comment.

RobertBColton commented on ENIGMAsystem/SHELL/Platforms/xlib/XLIBwindow.cpp:
```diff
@@ -267,15 +268,50 @@ bool window_get_stayontop() {
+    sh->min_width = sh->max_width = window_get_width();
+    sh->min_height = sh->max_height = window_get_height();
+  }
+  XSetWMSizeHints(disp, win, sh, XA_WM_NORMAL_HINTS);
```
```suggestion
  XSetWMNormalHints(disp, win, sh);
```

RobertBColton replied:
I guess it doesn't really save any lines, just a parameter, but you may as well do it anyway.

time-killer-games commented:
@RobertBColton please review again, I finally fixed what we were working on a day or two ago. On xlib, the window's draw area now properly scales with the window size, relative to the scaling GGS.

time-killer-games commented:
@JoshDreamland @RobertBColton Visual Studio really messed up the formatting. I'll use codeblocks next time to prevent this.

time-killer-games commented:
@RobertBColton @JoshDreamland this is now passing and ready for review (again). This fixes a ton of bugs in our win32 and xlib windowing systems, it's not even funny.

time-killer-games commented:
@RobertBColton one last bug i discovered, please don't merge this yet.

time-killer-games commented:
@RobertBColton ok i fixed that last bug there was. This should be good to go now. Josh wanted to talk to you about line endings so go ahead and do that when you can and hopefully we'll get this merged shortly after.

time-killer-games commented:
@JoshDreamland @RobertBColton I fixed borderless window switching on Linux to have the correct x/y position before and after switching that doesn't effect the client area's position on screen, which is just like how we have it in our Win32 and in GameMaker. The only problem is if you start the game in borderless mode, the first switch to bordered mode will change the positon of the client area on screen. I've concluded there is no workaround for this because there is no way to measure the position difference between the client area and border without the border being visible at least 1 time after the game has started. This is as good as it is going to get.

![GIFMaker org_XDrMVQ](https://user-images.githubusercontent.com/4379204/62075699-c2924780-b213-11e9-8b40-13025eee6790.gif)

RobertBColton commented on ENIGMAsystem/SHELL/Platforms/General/PFmain.cpp:
```diff
@@ -47,6 +48,8 @@ void set_program_args(int argc, char** argv) {
 }
 
 int enigma_main(int argc, char** argv) {
+  enigma::initGame = false;
```
Why initialize this here instead of the line where it's defined? Also, have you considered how this variable interacts with `game_restart`?

time-killer-games replied:
it does not get rest to false when game_restart() is called. This is intentional. We already have a gamestart variable in roomsystem.cpp; this is for an entirely different purpose, and the purpose is to apply windowing-related global game settings once and only once for each instance of the game running.

RobertBColton replied:
Actually, you don't need `initGame` at all, `gamestart` is already a paramter, you can use it and remove `initGame` now.
https://github.com/enigma-dev/enigma-dev/blob/d606db0007d3b9801e6290bd4271b1d24c7eb142/ENIGMAsystem/SHELL/Universal_System/roomsystem.cpp#L190

RobertBColton commented on ENIGMAsystem/SHELL/Platforms/Win32/WINDOWScallback.cpp:
```diff
@@ -18,228 +18,268 @@
-          lpMinMaxInfo->ptMinTrackSize.y = c.bottom-c.top;
-        }
-        break;
+namespace enigma {
```
I hate this several hundred lines of indentation and formatting changes. Yes, I like the 2 space indentation which is cleaner and more consistent with the rest of our actual style. However, it belongs in a separate pull request because it's too confusing what actually changed as far as logic in these source files. This is a general practice that I also do, and did just recently to format other source files. Please split the indentation and formatting changes out of this pull request.

time-killer-games commented:
@RobertBColton i tested gamestart and the value of it is reset to false when game_restart() is called. that's not what i need. that will break it. i'm not going to do that.

RobertBColton commented:
Oh, ok, I have another idea then. Again dump the initGame boolean, and do it in the initialization block. Then it will only be handled a single time per game at the start.

https://github.com/enigma-dev/enigma-dev/blob/7fcb1ce05d926daabdd76c38ac808da583f8b287/ENIGMAsystem/SHELL/Universal_System/Resources/loading.cpp#L128

RobertBColton commented on ENIGMAsystem/SHELL/Universal_System/roomsystem.cpp:
```diff
@@ -231,9 +231,19 @@ namespace enigma
-    // window sized by first room, can make visible now
-    enigma_user::window_set_visible(true);
+    if (!enigma::initGame) {
+      // GMS doesn't do this every room init; only game init
```
But now you've made it so the window is not recentered when switching rooms, which is GM8 behavior. Perhaps compare it to the latest GMS master.

time-killer-games replied:
GMStudio does not do what GM8 does in this regard. GM8 behavior in this case I feel YoYo made a good move when they dropped it. It's annoying to have the game window randomly move without the user's express permission. Plus, window_center() is broken on Xlib with multiple monitors, Josh wouldn't approve of GM8 behavior for that reason as well.

RobertBColton replied:
I can confirm I can only reproduce it in GM8.1 but not the latest GMSv1.4, so your decision is fine. And to clarify, GM Classic only did it when the next room had a different size. If the two rooms were the same size, it would not move the window.

RobertBColton commented on ENIGMAsystem/SHELL/Platforms/xlib/XLIBwindow.cpp:
```diff
@@ -45,8 +45,18 @@ Cursor NoCursor, DefCursor;
 namespace enigma {
 
+unsigned long decorationsPrevious = !showBorder;
+unsigned long decorationsCurrent = showBorder;
```
If these variables are not used in another source, they go to namespace anonymous. The `enigma::` namespace is only for internal variables used across compilation units.

```cpp
namespace {

} // anonymous namespace
```

time-killer-games commented:
switching sizeable property when the window is maximized to either the side, top/bottom, or corner of the screen results in messed up window decorations. I have no idea how to fix this. Our current code for checking whether the window is maximized only returns true if the window is maximized both horizontally and vertically. I've tried editing the code to window_get_maximized and split it up to return (maximized_horizontal || maximized_vertical) and that produced the same issues as the original code.

![Screenshot_20190730_130547](https://user-images.githubusercontent.com/4379204/62150382-e44f0580-b2cb-11e9-82c4-3516489b9669.png)

also, restoring the window in code from the maximized state to an unmaximized state currently does not do anything. whoever wrote that function for restoring the window either didn't test it or merged in an unfinished state.

time-killer-games commented:
@JoshDreamland are you ok with Robert merging this even though there is still this bug left I haven't been able to figure out yet mentioned here: https://github.com/enigma-dev/enigma-dev/pull/1831#issuecomment-516510779

time-killer-games commented:
@RobertBColton ready for mudge, if all parties involved still approve even after the two previous comments I posted.

RobertBColton commented:
Can you please try to move that initGame stuff into `initialize_everything` like I asked above?

time-killer-games commented:
> Can you please try to move that initGame stuff into `initialize_everything` and get rid of the boolean like I asked above? It's different from gamestart and will in fact only occur a single time, please try it.

I don't understand. I already did that?

RobertBColton commented:
No, you didn't.

time-killer-games commented:
Be more helpful than attempting to argue with me. Explain what I did wrong. I wrote the variable in the body of initialize_everything().

RobertBColton commented:
Ok, my problem is with initGame being global state basically when I don't think it needs to be. I was going to say move the window functions you call directly into `initialize_everything` but I guess the size of the first room still needs to be known. So, please just change the signature of the room function accept gameinit after gamestart as a parameter, and default it to false. Then remove `initGame` internal global.

time-killer-games commented:
doing it locally for testing first then i'll commit

time-killer-games commented:
@RobertBColton who the fuck turned off enigma bot? I'm looking for the signature for gotome and can't find it to save my life without it

time-killer-games commented:
@RobertBColton I tried it locally, doing exactly what you said, and the value is reset when the game restarts. 

in the roomsystem header:
`void gotome(bool gamestart = false, bool gameinit = false);`

in the roomsystem source:
    if (!gameinit) {
      // GMS doesn't do this every room init; only game init
      enigma_user::window_default(false);
      // window sized by first room, can make visible now
      enigma_user::window_set_visible(true);
      // required for global game setting resizeable window
      enigma_user::window_set_sizeable(enigma::isSizeable);
      // required for global game setting borderless window
      enigma_user::window_set_showborder(enigma::showBorder);
      // required for global game setting fullscreen window
      enigma_user::window_set_fullscreen(enigma::isFullScreen);
      gameinit = true;
    }

Therefore, please just merge it. You idea doesn't do what I needed specifically.

RobertBColton commented:
The idea does work, you have to also go to where it's called in initialize_everything and explicitly pass true for the second parameter.

time-killer-games commented:
I misunderstood.

time-killer-games commented:
gotome is not called in initialize_everything. with enigma bot turned off idk where it is. also whoever turned off enigma bot when i was in the middle of using it is going to get a knuckle sandwich and a creampie from me at the same time.

time-killer-games commented:
@RobertBColton it resets to false not just on game restart, but also every room start event

RobertBColton commented:
I'm sorry, it's in `game_start` which `initialize_everything` calls. However, I now see it won't work probably, since that's what `game_restart` calls.
https://github.com/enigma-dev/enigma-dev/blob/fd0923507e710657a54d11956467f4a04b3aeaf1/ENIGMAsystem/SHELL/Universal_System/roomsystem.cpp#L884

What you can do instead is just go back to what you had, remove `initGame` global again, and just make it a static inside `room::gotome`. Set it to false after you've run your initialization.
`static bool initGame = true;`

time-killer-games commented:
@RobertBColton didn't work. Same thing. The code executes every room start and game restart.

RobertBColton commented:
>The static keyword implies when initialized on the same line as the declaration the value can't be changed,

**_NO IT DOESN'T!_** Static has nothing to do with the constness of the local, it only affects the storage duration. That should work unless `game_restart` is really doing something tricky. Did you remove the other `initGame` while testing? If you didn't, this one will shadow the other one.

time-killer-games commented:
> > The static keyword implies when initialized on the same line as the declaration the value can't be changed,
> 
> **_NO IT DOESN'T!_** Static has nothing to do with the constness of the local, it only affects the storage duration. That should work unless `game_restart` is really doing something tricky. Did you remove the other `initGame` while testing? If you didn't, this one will shadow the other one.

i did remove the other one robert. come back to me when you ask josh about this and both me and him correct you again.

RobertBColton commented:
You're talking about **_const static_**, knock it off! And as far as I know, you haven't even taken programming 101, so stop insulting me. Everything you just said is false.

time-killer-games commented:
Would you like me to do a youtube video of a demonstration program, in which it will not change vale for the life of the program? give me ten minutes to install OBS studio

RobertBColton commented:
Don't waste more time on that, the video isn't going to help. Most of your behavior here is unacceptable. I only suggested the static local, which should be working, to avoid global state. I don't think the whole engine needs to be aware the game is being initialized, and it's a bad idea to just constantly create new global state in an application. It's harder to maintain and often unnecessary.

Consider reviewing your actions and behavior during this code review process. You exhibited a lot of impatience and I don't appreciate the insults. Approaching problems in this way does not lead to efficient and robust code, that's the point of the review process. I suggest you start taking a different approach to code reviews and view input and feedback as valuable to you, rather than a determent or a nuisance.

time-killer-games commented:
what is fact, is that your suggestion did not have the desired outcome. What else can you think of as an excuse for this to not be merged now? A global way to detect whether the first room has initialized once already is a very usseful global to have, and it will be useful for more things than it is being used for right now down the line.

RobertBColton commented:
It's a messy way to solve the problem compared to a more elegant way of solving it through control flow.

time-killer-games commented:
I just don't know of any other way to make initGame accessible to different various sources, which can be useful for a good number of reasons. If you have any ideas that actually do what you say they will, I'm opened to trying them. It's just hard to believe it'll pay off after trying about half a dozen things now that don't do what you said it would.

RobertBColton commented:
Since I agreed with you on changing the behavior to be more GMS like and not mess with the window when switching rooms. I was going to say I now find it acceptable to read the properties of the first room for the window size during `initialize_everything` which allows you to basically just move the entire bulk of your code over there. Just make sure it's after the rooms are loaded and there should be no problem and no need for any global or parameter.

time-killer-games commented:
@RobertBColton same result. the code executes every room start and game restart.

time-killer-games commented:
That works! Thank you! i will make the commits in a couple minutes.

time-killer-games commented:
@RobertBColton I'm really sorry for being toxic. Once it passes, I guess that means it can be merged.

time-killer-games commented:
@RobertBColton the DrawGUI event stopped working recently, I just noticed. Do you know what might be causing that?

RobertBColton commented on ENIGMAsystem/SHELL/Universal_System/Resources/loading.cpp:
```diff
@@ -124,14 +124,27 @@ namespace enigma
+    
+    // needed for DrawGUI event
+    enigma_user::screen_init();
+    enigma_user::screen_refresh();
```
Delete these two lines, they are in the room switch already. Move your code to execute before we go to the first room.

RobertBColton replied:
Another thing is, we will need the window visible much earlier in fact in the future. The reason? To render the game splash screen.

time-killer-games replied:
![Screenshot_20190730_230009](https://user-images.githubusercontent.com/4379204/62180409-f3f73a00-b31d-11e9-8fed-04334a7d0b0b.png)

time-killer-games replied:
notice the room size

time-killer-games replied:
Now would you just look at that:
![Screenshot_20190730_230043](https://user-images.githubusercontent.com/4379204/62180427-05d8dd00-b31e-11e9-836e-086241819f99.png)

time-killer-games replied:
Conclusion: ***STOP*** telling me to change shit for no fucking reason, only to discover what you told me to do breaks everything, and then jump around with a million different other ideas that don't work that you tell me to do for hours until by sheer luck you stumble across something that actually does what you wanted. I'm not your slave.

RobertBColton replied:
Right, so what we need to do now is pull the size of the first room during `initialize_everything`. Just include the room header that allows you to do `room_get_width(0);` and etc to initialize the window dimensions.

time-killer-games replied:
This could easily be solved 1 fucking way - don't review linux shit for however long you don't have a linux system. Let Josh be the one to do it, as he can actually test the stuff he wants done before asking me to add it, which you are in no position to do, and it is turning me into a person who simply does everything for you unpaid.

RobertBColton replied:
PS: I'm going to continue to delete comments that start new tangents. We're not going to do the splash screen in this pull request, and I have heard your opinions on the matter. The last I recalled, you were interested in such an _optional_ feature, but if not I won't even waste my time.

time-killer-games replied:
This is my precious time you are wasting. Either get a linux machine so you can review more competently and soberly, otherwise, leave it all to Josh.

RobertBColton replied:
Ok, I won't review anymore of your pull requests. Josh can take over now.

time-killer-games replied:
@JoshDreamland can you please mudge this puppy now

time-killer-games replied:
> Right, so what we need to do now is pull the size of the first room during `initialize_everything`. Just include the room header that allows you to do `room_get_width(0);` and etc to initialize the window dimensions. Might want to use room_first though, not sure which is safer/safest.

The problem with this is that it doesn't take the use of views into account.

RobertBColton replied:
Ah, that's correct, the views variables, which compute window size uses, haven't been loaded at this point.
https://github.com/enigma-dev/enigma-dev/blob/7fcb1ce05d926daabdd76c38ac808da583f8b287/ENIGMAsystem/SHELL/Universal_System/roomsystem.cpp#L224

I don't know then. You could split your code in two maybe, do all the initialize window settings in `initialize_everything` and then just handle initGame for the window size where you are now. That makes this problem a little trickier to solve.

Another way of going about it is to refactor compute window size to accept what room's view coordinates to use in the computation. Better yet, refactor compute window size to actually be a room function, so the window can ask for what size it should be given any room.

But now, since this is getting into a big refactor territory, I'm not really convinced in pushing you into doing it.

RobertBColton commented:
Looking a lot better, but you just need to tweak it a little more, move your code further up like I said. Also, what's up with the diff again, it looks like you accidentally recommitted those formatting changes.

RobertBColton commented:
Finally, I also want you to put a comment in `initialize_everything` so nobody reverts it back to GM8 behavior in the future. Put it near or above your code, but make it explicit that we're making this decision.

```cpp
// In pull request 1831, it was decided to adopt GMS behavior.
// The window is no longer moved, centered, or resized when switching rooms.
```

time-killer-games commented:
I fixed the formatting issue that was reverted somehow.

time-killer-games commented:
@RobertBColton I've cooked down. If it gets this merged sooner than Josh would, I'd like to do everything you want me to do to get this merged. Is it ok if you could give me a layout everything you want done? I promise I won't argue with you this time, and won't be upset if it doesn't work the first time.

time-killer-games commented:
@RobertBColton the changes you told me to do broke views in my key to success game. I reverted the code back to using the initGame in roomsystem.cpp and the problem is no longer there.

RobertBColton commented on ENIGMAsystem/SHELL/Platforms/Win32/WINDOWSwindow.cpp:
```diff
@@ -149,19 +149,22 @@ double window_get_alpha() {
 }
 
 void window_set_rectangle(int x, int y, int width, int height) {
+  if (window_get_fullscreen()) return;
```
None of this is correct, I tested these in GMS and GM8 at one point. The size that you set will be applied if the window is changed back to not fullscreen. The request can't be simply ignored.

time-killer-games replied:
This i did as a temporary fix because fullscreen breaks when you use these functions. The window will be be switched to borderless windowed mode rather than stay fullscreen, when these are called. No one should be attempting to resize the window unless they are in windowed mode, anyway.

time-killer-games replied:
but, what i can do, is save the values to windowX, windowY, windowWidth, and windowHeight, so that it will do exactly what you want still, now that i tink about it. Sound good?

time-killer-games replied:
@RobertBColton i tried removing the `if (window_get_fullscreen()) return;` and no matter what i do, attempting to use those functions to resize or reposition the window break both xlib and windows when used in fullscreen,

time-killer-games replied:
@RobertBColton are you going to address what I said here or are you just going to ignore it because you don't know how to create a better workaround for this than me.

time-killer-games replied:
Allowing these functions to work when fullscreen produces glitchy behavior on both windows and xlib. What do you want me to do? Are you just going to refuse to merge this because you prefer glitchy behavior over more correct behavior?

time-killer-games replied:
What GM obviously does, is save the window x/y/w/h to global state variables, and whatever these are set to, the game will go to when switching out of fullscreen. If we did excessive state checks like this in our main game loop like GM does, ENIGMA would suffer major performance with wasteful if statements called every step for every single last cached window state, among many other things GM does this with.

I thought you were against global state variables, anyway?

time-killer-games replied:
Is there any way this could be addressed in our follow up pr?

time-killer-games commented:
@RobertBColton @JoshDreamland I fixed the lag/stuttering and room view issues. I also fixed the borderless toggle so it doesn't change the client area positon on screen during the first switch to bordered mode if the game starts in borderless via global game settings, and I even paved the way for the optional splash screen to be added (shown on startup like GM).

Please review again when you can, thanks!

time-killer-games commented:
@JoshDreamland @RobertBColton I fixed the regressions. Please mudge.

time-killer-games commented:
I switched to the correct method to check for whether the current Desktop Environment is using KWin as the Window Manager; instead of checking if the value of the environment variable "XDG_CURRENT_DESKTOP" equals "KDE", I made it so that if the atom "KWIN_RUNNING" exists, then use KWin related code, otherwise use non-KWin related code. Where as the environment variable approach is unreliable and not all systems have XDG_CURRENT_DESKTOP defines, and not all K Desktop Environments are using KWin, and not all KWin Window Managers are using KDE. Now what we have is by far more correct.

time-killer-games commented:
@RobertBColton I deleted my discord. I would appreciate it to have some other means to contact you. Also, I've waited way too long for you to merge this. What is going on here?

time-killer-games commented:
@JoshDreamland Robert's not going to merge this ever it's looking like.

time-killer-games commented:
@RobertBColton @JoshDreamland ok, this time I have fully tested out this fucker on both Linux X11 and Win32. Everything is working beautifully. It would be really great if I got this merged today.

time-killer-games commented:
@JoshDreamland I have no way to contact Robert, and it is hard to tell whether he is still busy or just ignoring me here on GitGub at this point. Could you pass along how urgent it is he gets this merged? We have a game engine that has such broken windowing it is pretty much useless, this 1 pull request will clean up after all the shit fundies farted out his bum hole when he implemented SDL, and once this gets merged, we will have a real game engine again, or rather, perhaps for the very first time we will. Like, ever, have a real engine.

time-killer-games commented:
@RobertBColton All of the games and software I would like to create with ENIGMA, as well as all of my future contributions to the project, is permanently on hold until you merge this. It is 1 click away from being merged. Are you capable of something so very little? I doubt it.

time-killer-games commented:
@RobertBColton please merge. I've waited long enough.

JoshDreamland commented:
Robert is off attending commencement. But the project's pretty low-traffic, right now, so we may as well merge this and deal with any leftovers, later. The only problem I foresee is with the sheer number of window update calls after exiting fullscreen mode.

time-killer-games commented:
@JoshDreamland Now that this is merged, I'd like you to when you can to test this out and make sure I fixed window_center() on linux, it should be correct now, and the window will open in the center of your primary monitor, but if not, I need that fixed as my next top priority.
